src/main/java/com/google/cloud/pubsublite/flink/StopCondition.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.pubsublite.flink;
+
+import com.google.api.core.ApiFutures;
+import com.google.cloud.pubsublite.AdminClient;
+import com.google.cloud.pubsublite.Offset;
+import com.google.cloud.pubsublite.Partition;
+import com.google.cloud.pubsublite.SubscriptionPath;
+import com.google.cloud.pubsublite.TopicPath;
+import com.google.cloud.pubsublite.flink.internal.reader.PartitionFinishedCondition;
+import com.google.cloud.pubsublite.flink.internal.reader.PartitionFinishedCondition.Result;
+import com.google.cloud.pubsublite.internal.ExtractStatus;
+import com.google.cloud.pubsublite.internal.TopicStatsClient;
+import com.google.cloud.pubsublite.proto.Cursor;
+import com.google.common.collect.ImmutableMap;
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+public class StopCondition implements Serializable {
+  enum Condition {
+    INDEFINITE,
+    HEAD,
+    OFFSETS,
+  }
+
+  private final Condition condition;
+  private final Map<Partition, Offset> finalSplits;
```

Comment:
I would suggest having 3 subclasses of StopCondition for each condition case, and each handles those tofinishedCondition separately, also you would embed getHeadOffsets only in HEADStopCondition. I think that would be clearer.

Reply:
I like that, done